clang/lib/Sema/SemaBounds.cpp

```diff
@@ -1238,6 +1373,149 @@ namespace {
+            continue;
+          auto It = DstVars.begin();
+          for (; It != DstVars.end(); It++) {
+            if (EqualValue(S.Context, SrcV, *It, EquivExprs))
+              break;
+          }
+
+          // We searched all declared variables and found neither a constant nor
+          // a match for SrcV.
+          if (It == DstVars.end()) {
+            HasFreeVariables = true;
+            FreeVariables.push_back(SrcV);
+          }
+        }
+        return HasFreeVariables;
+      }
+
+      // Check free variables between E1 and E2. Append any found free variables
+      // to FreeVars with each free variable in E1 (resp. E2) having Pos1 (resp.
+      // Pos2).
+      static bool CheckFreeVarInExprs(Sema &S, Expr *E1, Expr *E2, FVPos Pos1,
+                                      FVPos Pos2, EquivExprSets *EquivExprs,
+                                      FreeVariableListTy &FreeVars) {
+        // Do not check free variables in bounds that contain indirect accesses.
+        if ((E1 && ReadsMemoryViaPointer(E1)) || (E2 && ReadsMemoryViaPointer(E2)))
+          return false;
+
+        bool HasFreeVariables = false;
+        EqualExprTy Vars1 = CollectVariableSet(S, E1);
+        EqualExprTy Vars2 = CollectVariableSet(S, E2);
+
+        if (AddFreeVariables(S, Vars1, Vars2, EquivExprs, Pos1, FreeVars))
+          HasFreeVariables = true;
+
+        if (AddFreeVariables(S, Vars2, Vars1, EquivExprs, Pos2, FreeVars))
+          HasFreeVariables = true;
+
+        return HasFreeVariables;
+      }
+
+      // AddFreeVariables maps each free variable in SrcVars w.r.t. DstVars
+      // to a pair <Variable, Pos>, and appends the pair to
+      // FreeVariablesWithPos.
+      static bool AddFreeVariables(Sema &S, const EqualExprTy &SrcVars,
+                                   const EqualExprTy &DstVars,
+                                   EquivExprSets *EquivExprs, FVPos Pos,
+                                   FreeVariableListTy &FreeVariablesWithPos) {
+        EqualExprTy FreeVariables;
+        if (GetFreeVariables(S, SrcVars, DstVars, EquivExprs, FreeVariables)) {
+          for (const auto V : FreeVariables) {
+            FreeVariablesWithPos.push_back(std::make_pair(V, Pos));
+          }
+          return true;
+        }
+        return false;
+      }
+
+      // IsEqualToConstant checks if Variable equals to some constant.
+      // Specifically, it checks if there is a set in EquivExprs that contains
+      // both Variable and a constant (i.e. IntegerType).
+      static bool IsEqualToConstant(Sema &S, Expr *Variable,
+                                    const EquivExprSets *EquivExprs) {
+        if (Variable->getType()->isPointerType() || !EquivExprs)
+          return false;
+
+        EqualExprTy EquivSet =
+            GetEqualExprSetContainingExpr(S, Variable, *EquivExprs);
+        for (const auto &E : EquivSet) {
+          if (E->getType()->isIntegerType())
```

Comment:
This function will return true if there is a set in `EquivExprs` that contains `Variable` and any other integer-typed expression (e.g. an integer-typed variable), not necessarily a constant. For example:

```
void f(int a, int b) {
    a = b;
}
```

`EquivExprs` contains the set `{ a, b }`. Since `b` has integer type, `IsEqualToConstant(a, EquivExprs)` will return true.

I suggest looking at the `Expr::isIntegerConstantExpr` method to test whether an expression is an integer constant.